zquery/shared/src/main/scala/zio/zquery/BlockedRequests.scala

```diff
@@ -0,0 +1,210 @@
+      extends BlockedRequests[R]
+
+  final case class Parallel[-R](left: BlockedRequests[R], right: BlockedRequests[R]) extends BlockedRequests[R]
+
+  final case class Sequential[-R](left: BlockedRequests[R], right: BlockedRequests[R]) extends BlockedRequests[R]
+
+  /**
+   * Flattens a collection of blocked requests to a sequence of sets of
+   * blocked requests, where each set represents requests that can be performed
+   * in parallel and sequential sets represent requests that must be performed
+   * sequentially.
+   */
+  private def flatten[R](blockedRequests: BlockedRequests[R]): List[Set[Single[R, _]]] = {
+
+    @tailrec
+    def loop[R](
+      blockedRequests: List[BlockedRequests[R]],
+      flattened: List[Set[Single[R, _]]]
+    ): List[Set[Single[R, _]]] = {
+      val (parallel, sequential) =
+        blockedRequests.foldLeft((Set.empty[Single[R, _]], List.empty[BlockedRequests[R]])) {
+          case ((parallel, sequential), blockedRequest) =>
+            val (set, seq) = step(blockedRequest)
+            (parallel ++ set, sequential ++ seq)
+        }
+      val updated = if (parallel.nonEmpty) parallel :: flattened else flattened
+      if (sequential.isEmpty) updated.reverse
+      else loop(sequential, updated)
+    }
+
+    loop(List(blockedRequests), List.empty)
+  }
+
+  /**
+   * Takes one step in evaluating a collection of blocked requests, returning a
+   * set of blocked requests that can be performed in parallel and a list of
+   * blocked requests that must be performed sequentially after those requests.
+   */
+  private def step[R](c: BlockedRequests[R]): (Set[Single[R, _]], List[BlockedRequests[R]]) = {
+
+    @tailrec
+    def loop[R](
+      blockedRequests: BlockedRequests[R],
+      stack: List[BlockedRequests[R]],
+      parallel: Set[Single[R, _]],
+      sequential: List[BlockedRequests[R]]
+    ): (Set[Single[R, _]], List[BlockedRequests[R]]) = blockedRequests match {
+      case Empty =>
+        if (stack.isEmpty) (parallel, sequential) else loop(stack.head, stack.tail, parallel, sequential)
+      case Sequential(left, right) =>
+        left match {
+          case Empty            => loop(right, stack, parallel, sequential)
+          case Sequential(l, r) => loop(Sequential(l, Sequential(r, right)), stack, parallel, sequential)
+          case Parallel(l, r) =>
+            loop(Parallel(Sequential(l, right), Sequential(r, right)), stack, parallel, sequential)
+          case o => loop(o, stack, parallel, right :: sequential)
+        }
+      case Parallel(left, right) => loop(left, right :: stack, parallel, sequential)
+      case o @ Single(_, _) =>
+        if (stack.isEmpty) (parallel ++ Set(o), sequential)
+        else loop(stack.head, stack.tail, parallel ++ Set(o), sequential)
+    }
+
+    loop(c, List.empty, Set.empty[Single[R, _]], List.empty)
+  }
+
+  def pipeline[R](blockedRequests: List[Set[Single[R, _]]]): List[List[Set[Single[R, _]]]] = {
```

Comment:
Some new types can help to communicate the information content of `List[List[Set[...]]]`, and may provide a place to put useful methods, as well.

Reply:
Yes, this file still needs some cleanup.